File: ios/Cove/Flows/CoinControlFlow/CoinControlContainer.swift
//
//  CoinControlContainer.swift
//  Cove
//
//  Created by Praveen Perera on 5/19/25.
//

import Foundation
import SwiftUI

struct CoinControlContainer: View {
    @Environment(AppManager.self) private var app

    let route: CoinControlRoute

    @State private var walletManager: WalletManager?
    @State private var manager: CoinControlManager?

    var body: some View {
        Group {
            if let walletManager, let manager {
                switch route {
                case .list:
                    UtxoListScreen(manager: manager)
                        .environment(walletManager)
                }
            } else {
                ProgressView()
                    .tint(.primary)
            }
        }
        .task(id: route.id()) {
            await loadManager()
        }
    }

    @MainActor
    private func loadManager() async {
        if walletManager != nil, manager != nil { return }

        do {
            Log.debug("Getting wallet for CoinControlRoute \(route.id())")
            let walletManager = try app.getWalletManager(id: route.id())
            let rustManager = try await walletManager.rust.newCoinControlManager()

            self.walletManager = walletManager
            manager = CoinControlManager(rustManager)
        } catch {
            handleCoinControlManagerError(error)
        }
    }

    private func handleCoinControlManagerError(_ error: Error) {
        switch error {
        case WalletManagerError.InitialScanIncomplete:
            app.showInitialScanIncompleteAlert()
            app.popRoute()
        case let WalletManagerError.DatabaseCorruption(walletId, errorMessage):
            Log.error("Wallet database corrupted for \(walletId): \(errorMessage)")
            app.alertState = TaggedItem(
                .walletDatabaseCorrupted(walletId: walletId, error: errorMessage)
            )
            app.popRoute()
        case WalletManagerError.WalletDoesNotExist:
            Log.error("Wallet does not exist for coin control route \(route)")
            app.alertState = .init(.general(
                title: "Wallet Not Found",
                message: "This wallet is no longer available."
            ))
            app.trySelectLatestOrNewWallet()
        case let walletError as WalletManagerError:
            Log.error("Unable to open wallet for coin control: \(walletError)")
            app.alertState = .init(.general(
                title: "Unable to Open Wallet",
                message: "The wallet could not be opened for coin control. Please try again from the wallet screen."
            ))
            app.popRoute()
        default:
            Log.error("Unable to open wallet for coin control: \(error)")
            app.alertState = .init(.general(
                title: "Unable to Open Wallet",
                message: "The wallet could not be opened for coin control. Please try again from the wallet screen."
            ))
            app.popRoute()
        }
    }
}

File: android/app/src/main/java/org/bitcoinppl/cove/flows/CoinControlFlow/CoinControlContainer.kt
package org.bitcoinppl.cove.flows.CoinControlFlow

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import org.bitcoinppl.cove.AppManager
import org.bitcoinppl.cove.CoinControlManager
import org.bitcoinppl.cove.TaggedItem
import org.bitcoinppl.cove.WalletManager
import org.bitcoinppl.cove_core.*
import org.bitcoinppl.cove_core.AppAlertState
import org.bitcoinppl.cove_core.types.*

/**
 * coin control container - manages WalletManager + CoinControlManager lifecycle
 * ported from iOS CoinControlContainer.swift
 */
@Composable
fun CoinControlContainer(
    app: AppManager,
    route: CoinControlRoute,
    modifier: Modifier = Modifier,
) {
    // extract wallet ID from route
    val walletId =
        when (route) {
            is CoinControlRoute.List -> route.v1
        }

    var walletManager by remember(walletId) { mutableStateOf<WalletManager?>(null) }
    var manager by remember(walletId) { mutableStateOf<CoinControlManager?>(null) }
    val tag = "CoinControlContainer"

    // async initialize managers
    LaunchedEffect(walletId) {
        try {
            android.util.Log.d(tag, "getting wallet for CoinControlRoute $walletId")

            val wm = app.getWalletManager(walletId)
            val rustManager = wm.rust.newCoinControlManager()
            val ccm = CoinControlManager(rustManager)

            walletManager = wm
            manager = ccm
        } catch (e: WalletManagerException.InitialScanIncomplete) {
            android.util.Log.e(tag, "initial scan incomplete", e)
            app.alertState =
                TaggedItem(
                    AppAlertState.General(
                        title = "Initial Scan Incomplete",
                        message = "Can't send until initial scan completes.",
                    ),
                )
            app.popRoute()
        } catch (e: WalletManagerException.DatabaseCorruption) {
            android.util.Log.e(tag, "wallet database corrupted for ${e.`id`}: ${e.`error`}", e)
            app.alertState =
                TaggedItem(
                    AppAlertState.WalletDatabaseCorrupted(walletId = e.`id`, error = e.`error`),
                )
            app.popRoute()
        } catch (e: WalletManagerException.WalletDoesNotExist) {
            android.util.Log.e(tag, "wallet does not exist for CoinControlRoute $walletId", e)
            app.alertState =
                TaggedItem(
                    AppAlertState.General(
                        title = "Wallet Not Found",
                        message = "This wallet is no longer available.",
                    ),
                )
            app.trySelectLatestOrNewWallet()
        } catch (e: WalletManagerException) {
            android.util.Log.e(tag, "unable to open wallet for coin control", e)
            app.alertState =
                TaggedItem(
                    AppAlertState.General(
                        title = "Unable to Open Wallet",
                        message = "The wallet could not be opened for coin control. Please try again from the wallet screen.",
                    ),
                )
            app.popRoute()
        } catch (e: Exception) {
            android.util.Log.e(tag, "unable to initialize coin control", e)
            app.alertState =
                TaggedItem(
                    AppAlertState.General(
                        title = "Unable to Open Wallet",
                        message = "The wallet could not be opened for coin control. Please try again from the wallet screen.",
                    ),
                )
            app.popRoute()
        }
    }

    // cleanup on disappear or wallet change
    DisposableEffect(walletId) {
        onDispose {
            manager?.close()
        }
    }

    // render
    val currentManager = manager
    when {
        walletManager != null && currentManager != null -> {
            when (route) {
                is CoinControlRoute.List -> {
                    UtxoListScreen(
                        manager = currentManager,
                        app = app,
                        modifier = modifier,
                    )
                }
            }
        }
        else -> {
            Box(
                modifier = modifier.fillMaxSize(),
                contentAlignment = Alignment.Center,
            ) {
                CircularProgressIndicator()
            }
        }
    }
}
